chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/pendingListing/reviewer/duplicate/TestToolDuplicateReviewer.java

```diff
@@ -0,0 +1,68 @@
+package gov.healthit.chpl.validation.pendingListing.reviewer.duplicate;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiPredicate;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import gov.healthit.chpl.dto.PendingCertificationResultDTO;
+import gov.healthit.chpl.dto.PendingCertificationResultTestToolDTO;
+import gov.healthit.chpl.dto.PendingCertifiedProductDTO;
+import gov.healthit.chpl.util.ErrorMessageUtil;
+
+@Component("testToolDuplicateReviewer")
+public class TestToolDuplicateReviewer {
+    private ErrorMessageUtil errorMessageUtil;
+
+    @Autowired
+    public TestToolDuplicateReviewer(final ErrorMessageUtil errorMessageUtil) {
+        this.errorMessageUtil = errorMessageUtil;
+    }
+
+    public void review(final PendingCertifiedProductDTO listing, final PendingCertificationResultDTO certificationResult) {
+
+        DuplicateReviewResult<PendingCertificationResultTestToolDTO> testToolDuplicateResults =
+                new DuplicateReviewResult<PendingCertificationResultTestToolDTO>(getPredicate());
+
+        if (certificationResult.getTestTools() != null) {
+            for (PendingCertificationResultTestToolDTO dto : certificationResult.getTestTools()) {
+                testToolDuplicateResults.addObject(dto);
+            }
+        }
+
+        if (testToolDuplicateResults.duplicatesExist()) {
+            listing.getWarningMessages().addAll(
+                    getWarnings(testToolDuplicateResults.getDuplicateList(), certificationResult.getNumber()));
+            certificationResult.setTestTools(testToolDuplicateResults.getUniqueList());
+        }
+    }
+
+    private List<String> getWarnings(final List<PendingCertificationResultTestToolDTO> duplicates, final String criteria) {
+        List<String> warnings = new ArrayList<String>();
+        for (PendingCertificationResultTestToolDTO duplicate : duplicates) {
+            String warning = errorMessageUtil.getMessage("listing.criteria.duplicateTestTool",
+                    criteria, duplicate.getName(), duplicate.getVersion());
+            warnings.add(warning);
+        }
+        return warnings;
+    }
+
+    private BiPredicate<PendingCertificationResultTestToolDTO, PendingCertificationResultTestToolDTO> getPredicate() {
+        return new BiPredicate<PendingCertificationResultTestToolDTO, PendingCertificationResultTestToolDTO>() {
+            @Override
+            public boolean test(final PendingCertificationResultTestToolDTO dto1,
+                    final PendingCertificationResultTestToolDTO dto2) {
+                if (dto1.getName() != null && dto2.getName() != null
```

Comment:
These are all a bunch of `&&`,s right? So it's:
```
if (a & b & c & d)
return e & f
else
return false
```
Why not
```
 return (a & b & c & d & e & f)
```
where each of a, b, c, ... are the clauses you have?

Reply:
I typically leave the null checks separate from the return value simply for readability, but it doesn't really matter to me.

What would be your thoughts on using this Java 7 Objects.equals method.  https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#equals(java.lang.Object,%20java.lang.Object)

I need to make sure the logic is the same, but thought it was worth a look.

Reply:
It looks like the Object.equals(a, b), when neither a nor b are null, simply calls a.equals(b). If that's the case then I think that could work here, but we'd probably need to make sure our "a.equals" override, if it exists, is correct, and if it doesn't, works. Also, Object.equals(a, b), for both a and b as null, would return true. Would we want that here? If we can guarantee that at least one of a or b is non-null, and the a.equals() method works, then I think that would be okay here, but I can't tell that from here

Reply:
I generally very much like the idea of overriding equals (and hashCode). The only thing I would worry about in this case is are there times when just the test tool name is filled in or just the test tool id? The answer might be no, but just make sure you test various cases of a user uploading a file with a bad test tool name, good test tool name, and then when a user goes through the confirm dialog and adds a test tool are the id and name both filled in or just the id? I think we ran into something similar with the macra measures recently where the name or id was null in one use case but not another. Also just want to make sure there are no unintended consequences when overriding equals. It is possible that method gets called on objects a very low level so if for instance we have a Set<PendingCertifiedProductTestToolDTO> anywhere, the equals() behavior could affect that. All that to say - I like the idea, go for it, be careful!